internal/topology/variables/cluster_variable_validation_test.go

```diff
@@ -0,0 +1,322 @@
+					},
+				},
+				{
+					Name: "zone",
+					Value: apiextensionsv1.JSON{
+						Raw: []byte(`"longerThanOneCharacter"`),
+					},
+				},
+				{
+					Name: "location",
+					Value: apiextensionsv1.JSON{
+						Raw: []byte(`"us-east-1"`),
+					},
+				},
+			},
+		},
+		{
+			name: "Error if required ClusterClassVariable is not defined in ClusterVariables.",
+			clusterClassVariables: []clusterv1.ClusterClassVariable{
+				{
+					Name:     "cpu",
+					Required: true,
+					Schema: clusterv1.VariableSchema{
+						OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+							Type:    "integer",
+							Minimum: pointer.Int64(1),
+						},
+					},
+				},
+				{
+					Name:     "zone",
+					Required: true,
+					Schema: clusterv1.VariableSchema{
+						OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+							Type:      "string",
+							MinLength: pointer.Int64(1),
+						},
+					},
+				},
+			},
+
+			clusterVariables: []clusterv1.ClusterVariable{
+				// cpu is missing in the ClusterVariables but is required in ClusterClassVariables.
+				{
+					Name: "zone",
+					Value: apiextensionsv1.JSON{
+						Raw: []byte(`"longerThanOneCharacter"`),
+					},
+				},
+			},
+			wantErr: true,
+		},
+		{
+			name: "Error if ClusterVariable defined which has no ClusterClassVariable definition.",
+			clusterClassVariables: []clusterv1.ClusterClassVariable{
+				{
+					Name:     "cpu",
+					Required: true,
+					Schema: clusterv1.VariableSchema{
+						OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+							Type:    "integer",
+							Minimum: pointer.Int64(1),
+						},
+					},
+				},
+				{
+					Name:     "zone",
+					Required: true,
+					Schema: clusterv1.VariableSchema{
+						OpenAPIV3Schema: clusterv1.JSONSchemaProps{
+							Type:      "string",
+							MinLength: pointer.Int64(1),
+						},
+					},
+				},
+			},
+
+			clusterVariables: []clusterv1.ClusterVariable{
+				// location is defined here but not in the ClusterClassVariables
+				{
+					Name: "location",
+					Value: apiextensionsv1.JSON{
+						Raw: []byte(`"us-east-1"`),
+					},
+				},
+				{
+					Name: "cpu",
+					Value: apiextensionsv1.JSON{
+						Raw: []byte(`1`),
+					},
+				},
+
+				{
+					Name: "zone",
+					Value: apiextensionsv1.JSON{
+						Raw: []byte(`"longerThanOneCharacter"`),
+					},
+				},
+			},
+			wantErr: true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			errList := ValidateClusterVariables(tt.clusterVariables, tt.clusterClassVariables,
+				field.NewPath("spec", "topology", "variables"))
+
+			if tt.wantErr {
+				g.Expect(errList).NotTo(BeEmpty())
+				return
+			}
+			g.Expect(errList).To(BeEmpty())
+		})
+	}
+}
+
+func Test_ValidateClusterVariable(t *testing.T) {
+	tests := []struct {
+		name                 string
+		clusterClassVariable *clusterv1.ClusterClassVariable
+		clusterVariable      *clusterv1.ClusterVariable
+		wantErr              bool
+	}{
+		{
+			name: "Valid integer",
```

Comment:
Can we add ~2 cases for Invalid integer (min & max) and the same for Invalid string (one of the formats / Pattern match)?